app/downloader.py

```diff
@@ -0,0 +1,237 @@
+        Retry.DEFAULT_BACKOFF_MAX = self.BACKOFF_MAX
+        retry_strategy = Retry(
+            total=self.RETRY_COUNT,
+            raise_on_status=True,
+            backoff_factor=self.BACKOFF_FACTOR,
+            # retry on common server side errors and non-critical client side errors
+            status_forcelist={413, 429, 500, 502, 503, 504},
+            allowed_methods=["GET"],
+        )
+        adapter = HTTPAdapter(
+            pool_connections=self.MAX_CONCURRENT_DOWNLOAD,
+            pool_maxsize=self.MAX_CONCURRENT_DOWNLOAD,
+            max_retries=retry_strategy,
+        )
+        session.mount("https://", adapter)
+        session.mount("http://", adapter)
+
+        # cleanup proxy if any
+        proxies = {"http": "", "https": ""}
+        session.proxies.update(proxies)
+
+        # register the connection pool
+        self.session = session
+        self._proxy_set = False
+
+    def shutdown(self):
+        self.session.close()
+
+    def configure_proxy(self, proxy: str):
+        # configure proxy
+        self._proxy_set = True
+        proxies = {"http": proxy, "https": ""}
+        self.session.proxies.update(proxies)
+
+    def cleanup_proxy(self):
+        self._proxy_set = False
+        self.configure_proxy("")
+
+    def _path_to_url(self, base: str, path: Union[Path, str]) -> str:
+        # regulate base url, add suffix / to it if not existed
+        base = f"{base}/" if not base.endswith("/") else base
+        # convert to Path if path is str
+        path = Path(path) if isinstance(path, str) else path
+
+        relative_path = path
+        # if the path is relative to /
+        try:
+            relative_path = path.relative_to("/")
+        except ValueError:
+            pass
+
+        quoted_path = quote_from_bytes(bytes(relative_path))
+
+        # switch scheme if needed
+        _url_parsed = urlparse(urljoin(base, quoted_path))
+        # unconditionally set scheme to HTTP if proxy is applied
+        if self._proxy_set:
+            _url_parsed = _url_parsed._replace(scheme="http")
+
+        return _url_parsed.geturl()
+
+    @partial(_retry, RETRY_COUNT, OUTER_BACKOFF_FACTOR, BACKOFF_MAX)
+    def download(
+        self,
+        path: Union[Path, str],
+        dst: Union[Path, str],
+        digest: Optional[str] = None,
+        *,
+        url_base: str,
+        cookies: Optional[Dict[str, str]] = None,
+        headers: Optional[Dict[str, str]] = None,
+    ) -> int:
+        url = self._path_to_url(url_base, path)
+        cookies = cookies if cookies else {}
+        headers = headers if headers else {}
+        dst = Path(dst)
+
+        # specially deal with empty file
+        if digest == self.EMPTY_STR_SHA256:
+            dst.write_bytes(b"")
+            return 0
+
+        try:
+            error_count = 0
+            response = self.session.get(
+                url, stream=True, cookies=cookies, headers=headers
+            )
+
```

Comment:
`response.raise_for_status()` is removed. Is it OK?

Reply:
Yes, actually this line is never used as the exception is raised by the `urllib3.Retry` module.

The `raise_on_status` param is set when initializing the `Retry` adapter:
```python
        retry_strategy = Retry(
            total=self.RETRY_COUNT,
            raise_on_status=True,
            backoff_factor=self.BACKOFF_FACTOR,
            # retry on common server side errors and non-critical client side errors
            status_forcelist={413, 429, 500, 502, 503, 504},
            allowed_methods=["GET"],
        )
```

Reply:
Let me confirm. The document says:
```
raise_on_status ([bool](https://docs.python.org/3/library/functions.html#bool)) – Similar meaning to raise_on_redirect: whether we should raise an exception, or return a response, if status falls in status_forcelist range and retries have been exhausted.
```
So if 400 is returned as a status code, status doesn't fall in `status_forecelist`, so response.raise_for_status() should  not be removed.
Please confirm. I am sorry if my understanding is wrong :-)